### PR TITLE
Respect unnormalized logits input in Categorical/Multinomial dist

### PR DESCRIPTION
**Pull request by fehiepsi**
Currently, logits of these distributions are normalized. The changes in this PR have the following folds
+ Removes the unexpected behaviour for users when `logits` input values are modified at __init__ 
+ Removes the unnecessary logsumexp op when only sampling is necessary
+ Matches with `tensorflow` definition of `logits` ([ref](https://www.tensorflow.org/probability/api_docs/python/tfp/distributions/Multinomial#args))
+ (personally) allow me inject a custom log_prob term to joint density (by setting value=0)

This PR also fixes the bug of missing `axis=-1` for `logsumexp` in previous PRs.

### CODE REVIEW

**Comment by fritzo** [comment]
Interesting, could we have a deeper design discussion?

I don't fully understand this PR, but I do like the general idea of allowing the partition function `Z` to pass through distribution methods (by avoiding normalization). In Funsor we've made the decision to keep low-level funsors non-normalized but to keep the minipyro distributions interface normalized. Is that similar to what you're doing in this PR, to keep low-level functions non-normalized while ensuring `dist.Categorical(logits=logits).log_prob(-)` is indeed normalized?

**Comment by fehiepsi** [comment]
When we have a `custom_log_prob`, with `total_count=1`, `value=0`, the normalize term `total_count * logsumexp(self.logits, axis=-1) - gammaln(total_count + 1)` will be `custom_log_prob` and the remaining term `np.sum(value * self.logits - gammaln(value + 1), axis=-1)` will be `0`. So the Multinomial log_prob will be `custom_log_prob`.

The BernoulliLogits only works for the case `custom_log_prob << 0`, where `binary_entropy_with_logits(logits, 1.) ~ logits`. Bernoulli trick does not work when `custom_log_prob ~0` or completely wrong when `custom_log_prob > 0` (in case we calculate unnormalized log_prob).

**Comment by fehiepsi** [comment]
> Is that similar to what you're doing in this PR, to keep low-level functions non-normalized while ensuring dist.Categorical(logits=logits).log_prob(-) is indeed normalized?

I guess it is a bit different. The distributions `Categorical` and `Multinomial` are normalized. But we can inject an unnormalized term (calculated outside these distributions) by using the above trick. By using the **invalid** sample `value=0` (not belongs to the support), the custom log prob term will not be normalized. I think that the logic of `funsor` is correct, while here it is just a trick. ^^!

**Comment by neerajprad** [comment]
@fritzo - This is mostly just a cleanup/bugfix PR. This allows for an additional hack, IIUC from what @fehiepsi mentioned, of injecting some custom term into the log joint computation by using `Multinomial(total_count=1, logits=-custom_term).log_prob(value=0) --> custom_term`. I'm calling it a hack because it will throw an error with validation check enabled. :) I think @fehiepsi has some models where this trick be useful, and it will be interesting to see if we can use funsors as the backend for such models.

**Comment by neerajprad** [comment]
Merging this since this is just a bugfix PR, but we should have a separate design discussion on "allowing the partition function Z to pass through distribution methods (by avoiding normalization)" which will be useful for models that can exploit conjugacy, for instance.

**Comment by fehiepsi** [comment]
> using Multinomial(total_count=1, logits=-custom_term).log_prob(value=0) --> custom_term

Opps... thanks for correcting me (about the sign of the custom term).
